### PR TITLE
MPI fixes (macOS build/link flags), oneAPI ifort->ifx switch; turn off MSMPI CI

### LINKED ISSUE #975: Latest Xcode update has broken MPI builds on macOS
### Description

`mpif90 -showme:compile` requires these flags
`-I/opt/homebrew/Cellar/open-mpi/4.1.5/include -Wl,-flat_namespace -Wl,-commons,use_dylibs -I/opt/homebrew/Cellar/open-mpi/4.1.5/lib`

But the `commons` flag is not recognized by `ld` anymore. For example, with the example MPI packags

```
federico@Federicos-MBP metapackage_mpi % fpm build
 + which mpiexec
/opt/homebrew/bin/mpiexec
test_mpi                               failed.
[100%] Compiling...
ld: unknown options: -commons -commons
collect2: error: ld returned 1 exit status
<ERROR> Compilation failed for object " test_mpi "
<ERROR> stopping due to failed compilation
STOP 1
```

This is likely the same issue as https://github.com/pmodels/mpich/issues/6684 [Intel forum](https://community.intel.com/t5/Intel-Fortran-Compiler/Mac-Xcode-15-0-unknown-options-commons/td-p/1526357)

I fear the only way to solve this issue is to conditionally remove those flags based on the current OS and MPI version, not exactly nice, at least until a new OpenMPI homebrew release will fix this.

### Expected Behaviour

Should just work

### Version of fpm

0.9.0, alpha

### Platform and Architecture

MacOS 14, homebrew OpenMPI 4.1.5

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by perazz**
- Fix #975
- Add %MSMPI_BIN% with Git-Bash style in CI
- CI: replace `ifort` with `ifx` in oneAPI runner. Fpm build now fails [due to this issue](https://fortran-lang.discourse.group/t/use-of-uninitialized-value-warning-with-check-all-option-of-ifx/6823/4)
- Fix Intel fpm build by using `-check all,nouninit` (see above)
- Fix missing ifx,icx,icpx compilers from the current Intel MPI wrappers (https://github.com/easybuilders/easybuild-framework/issues/4056)

@fortran-lang/admins

### CODE REVIEW

File: src/fpm_meta.f90 [hunk 1]
```diff
@@ -1598,4 +1600,124 @@ type(string_t) function mpi_wrapper_query(mpilib,wrapper,command,verbose,error)
...
+
+   select case (compiler%id)
+      case (id_intel_classic_windows,id_intel_llvm_windows)
+          is_link_argument = string=='/link' &
+                             .or. str_begins_with_str(string,'/LIBPATH')&
+                             .or. str_ends_with(string,'.lib') ! always .lib whether static or dynamic
+      case default
+
+          ! fix OpenMPI's Fortran wrapper bug (https://github.com/open-mpi/ompi/issues/11636) here
+          is_link_argument = (    str_begins_with_str(string,'-L') &
```
Comment (ivan-pi) [inline src/fpm_meta.f90:1616]: Should this logic be invoked only for OpenMPI wrappers in a certain version range? Or would is the extra effort unwarranted?

Comment (perazz) [reply src/fpm_meta.f90:1616]: Here, the [meson](https://github.com/mesonbuild/meson/blob/ebf5757c59d9bceb953dc531c757e07dcf83c081/mesonbuild/dependencies/mpi.py#L134C1-L135C1) approach is replicated, i.e., to remove all flags that are not strictly 1) library directories 2) library flags 3) important warning flags. It probably shielded it from the macOS issue above, and may as well do the same in the future

**Comment by perazz** [comment]
It seems like something is now failing in the oneAPI CI (classic intel compilers deprecated?). I see that no fpm actions have been run for the past few months, so that might be the case. I will do some more tests

**Review by henilp105 — APPROVED** [review]
Thanks @perazz , Looks Good to me.

**Review by gnikit — COMMENTED** [review]
Looks good, there are quite a few changes though. I left some comments inline mostly with clean-up, underlying code seems good.

To make review faster, when possible, could we try and stick with 1 fix/feature per PR. One of the reasons why I delayed reviewing this is because fixes kept being added in this PR.

File: src/fpm_environment.f90 [hunk 2]
```diff
@@ -50,6 +50,8 @@ integer function get_os_type() result(r)
 
         first_run = .false.
         r = OS_UNKNOWN
+        length = 0
+        rc = 0
```
[the remark this answers is not in the document]
Comment (perazz) [reply src/fpm_environment.f90]: Thanks for catching, these are leftovers from before I introduced a [fix to this](https://fortran-lang.discourse.group/t/use-of-uninitialized-value-warning-with-check-all-option-of-ifx/6823/4)

File: src/fpm_meta.f90 [hunk 3]
```diff
@@ -487,6 +490,9 @@ subroutine init_mpi(this,compiler,error)
 
     !> Cleanup
     call destroy(this)
+    fwrap = string_t("")
+    cwrap = string_t("")
+    cxxwrap = string_t("")
```
[the remark this answers is not in the document]
Comment (perazz) [reply src/fpm_meta.f90]: see above

**Comment by perazz** [comment]
> One of the reasons why I delayed reviewing this is because fixes kept being added in this PR.

I agree, committed to it. Unfortunately new issues kept arising in the CI as I was fixing the macOS flags

**Review by gnikit — APPROVED** [review]
LGTM :rocket:

**Comment by perazz** [comment]
Thank you @gnikit @henilp105 @ivan-pi for the reviews, let's merge in a few days if there are no more comments.

**Comment by gnikit** [comment]
I will have a look in the windows metapackages CI to see if we can get it to work more reliably. If the Action is not robust yet, it might be worth disabling  whilst troubleshooting to merge in main.

**Pull request merged**